dhis-2/dhis-api/src/test/java/org/hisp/dhis/dataApproval/DataApprovalWorkflowTest.java

```diff
@@ -0,0 +1,168 @@
+package org.hisp.dhis.dataApproval;
+
+/*
+ * Copyright (c) 2004-2020, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.dataapproval.DataApprovalLevel;
+import org.hisp.dhis.dataapproval.DataApprovalWorkflow;
+import org.hisp.dhis.dataset.DataSet;
+import org.hisp.dhis.period.*;
+import org.junit.Test;
+
+import java.util.List;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author Jim Grace
+ */
+public class DataApprovalWorkflowTest
+{
+    @Test
+    public void testGetSortedLevels()
+    {
+        DataApprovalLevel level1 = new DataApprovalLevel( "level1", 1, null );
+        level1.setLevel( 1 );
+
+        DataApprovalLevel level2 = new DataApprovalLevel( "level2", 2, null );
+        level2.setLevel( 2 );
+
+        DataApprovalLevel level3 = new DataApprovalLevel( "level3", 3, null );
+        level3.setLevel( 3 );
+
+        DataApprovalWorkflow workflow = new DataApprovalWorkflow( "test workflow", new DailyPeriodType(), newHashSet(level3, level2, level1 ) );
+
+        List<DataApprovalLevel> levels = workflow.getSortedLevels();
+
+        assertEquals( 1, levels.get( 0 ).getLevel() );
+        assertEquals( 2, levels.get( 1 ).getLevel() );
+        assertEquals( 3, levels.get( 2 ).getLevel() );
+    }
+
+    @Test
+    public void testGetWorkflowSqlCoEnddateExtension()
```

Comment:
This test is perfectly fine, but you may also consider using Junit parametrized test: https://github.com/junit-team/junit4/wiki/Parameterized-tests

Reply:
Thanks for the tip. There are undoubtedly places where I should be using parameterized JUnit tests. One reason to keep this test as coded, however, is because it's also testing to be sure that the longest extension is always returned as data sets are added to the workflow. Note that the (nominal period length) * (the number of periods) is always increasing, so the latest one should always be returned. (It might be worth mentioning this in a method JavaDoc header!)